src/MongoDB.Driver.Core/Core/Connections/BinaryConnection.cs

```diff
@@ -769,9 +793,7 @@ public ResponseMessage DecodeMessage(IByteBuffer buffer, IMessageEncoderSelector
                 _stopwatch.Restart();
                 using (var stream = new ByteBufferStream(buffer, ownsBuffer: false))
                 {
-                    var encoderFactory = new BinaryMessageEncoderFactory(stream, _messageEncoderSettings);
-                    var encoder = encoderSelector.GetEncoder(encoderFactory);
-                    message = (ResponseMessage)encoder.ReadMessage();
+                    message = ReadResponseMessage(stream, encoderSelector);
```

Comment:
Instead of the new ReadResponseMessage helper method, refactor like this instead:

                    var encoderFactory = new BinaryMessageEncoderFactory(stream, _messageEncoderSettings, _compressorSource);

                    var opcode = PeekOpcode(stream);
                    if (opcode == Opcode.Compressed)
                    {
                        var compresedMessageEncoder = encoderFactory.GetCompressedMessageEncoder(encoderSelector);
                        var compressedMessage = (CompressedMessage)compresedMessageEncoder.ReadMessage();
                        message = (ResponseMessage)compressedMessage.OriginalMessage;
                    }
                    else
                    {
                        var encoder = encoderSelector.GetEncoder(encoderFactory);
                        message = (ResponseMessage)encoder.ReadMessage();
                    }

If the opcode is not `Compressed` we execute the exact same code path as before.

Reply:
private Opcode PeekOpcode(BsonStream stream)
            {
                var savedPosition = stream.Position;
                stream.Position += 12;
                var opcode = (Opcode)stream.ReadInt32();
                stream.Position = savedPosition;
                return opcode;
            }